File: AkkaSTMExample/src/main/java/org/akka/essentials/stm/pingpong/example2/Game.java
package org.akka.essentials.stm.pingpong.example2;

import akka.actor.ActorSystem;
import scala.concurrent.stm.Ref;
import scala.concurrent.stm.japi.STM;

public class Game {
	public static void main(String args[]) {

		ActorSystem _system = ActorSystem.create("Agent-example");
		Ref.View<String> turn = STM.newRef("");
		PingPong table = new PingPong(turn);

		Thread alice = new Thread(new Player("bob", table));
		Thread bob = new Thread(new Player("alice", table));

		alice.setName("alice");
		bob.setName("bob");
		alice.start(); // alice starts playing
		bob.start(); // bob starts playing
		try {
			// Wait 5 seconds
			Thread.sleep(5000);
		} catch (InterruptedException e) {
		}

		table.hit("DONE"); // cause the players to quit their threads.
		try {
			Thread.sleep(3000);
		} catch (InterruptedException e) {
		}
		_system.shutdown();
	}
}

File: AkkaSTMExample/src/main/scala/org/akka/essentials/stm/pingpong/example2/Game.scala
package org.akka.essentials.stm.pingpong.example2

import scala.concurrent.stm.Ref

import akka.actor.ActorSystem

object Game {

  def main(args: Array[String]): Unit = {

    val _system = ActorSystem("Ref-example")
    val turn = Ref(new String())
    val table = new PingPong(turn)

    val alice = new Thread(new Player("bob", table))
    val bob = new Thread(new Player("alice", table))

    alice.setName("alice")
    bob.setName("bob")
    alice.start // alice starts playing
    bob.start // bob starts playing
    try {
      // Wait .5 seconds
      Thread.sleep(500)
    } catch {
      case _ =>
      // eat the exception
    }
    table.hit("DONE") // cause the players to quit their threads.
    try {
      Thread.sleep(1000)
    } catch {
      case _ =>
      // eat the exception
    }
    _system.shutdown
  }

}